File: src/main/java/net/flyingfishflash/loremlist/core/configuration/WebSecurityConfiguration.java
package net.flyingfishflash.loremlist.core.configuration;

import java.util.List;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.HttpMethod;
import org.springframework.security.config.Customizer;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
import org.springframework.security.config.http.SessionCreationPolicy;
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationConverter;
import org.springframework.security.oauth2.server.resource.authentication.JwtGrantedAuthoritiesConverter;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.web.cors.CorsConfiguration;
import org.springframework.web.cors.CorsConfigurationSource;
import org.springframework.web.cors.UrlBasedCorsConfigurationSource;

@Configuration
@EnableWebSecurity
public class WebSecurityConfiguration {

  @Bean
  public CorsConfigurationSource corsConfigurationSource() {
    CorsConfiguration configuration = new CorsConfiguration();
    configuration.setAllowedOriginPatterns(List.of("*"));
    configuration.setAllowedMethods(List.of("*"));
    configuration.setAllowedHeaders(List.of("*"));
    configuration.setAllowCredentials(true);
    configuration.setExposedHeaders(List.of("X-Auth-Token", "Authorization"));
    UrlBasedCorsConfigurationSource source = new UrlBasedCorsConfigurationSource();
    source.registerCorsConfiguration("/**", configuration);
    return source;
  }

  @Bean
  public JwtAuthenticationConverter jwtAuthenticationConverter() {
    JwtGrantedAuthoritiesConverter grantedAuthoritiesConverter =
        new JwtGrantedAuthoritiesConverter();
    grantedAuthoritiesConverter.setAuthoritiesClaimName("grants");

    JwtAuthenticationConverter jwtAuthenticationConverter = new JwtAuthenticationConverter();
    jwtAuthenticationConverter.setJwtGrantedAuthoritiesConverter(grantedAuthoritiesConverter);
    return jwtAuthenticationConverter;
  }

  @Bean
  public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
    http.cors(Customizer.withDefaults())
        .csrf(AbstractHttpConfigurer::disable)
        .oauth2ResourceServer(oauth2 -> oauth2.jwt(Customizer.withDefaults()))
        .sessionManagement(
            session -> session.sessionCreationPolicy(SessionCreationPolicy.STATELESS))
        .authorizeHttpRequests(
            auth ->
                auth.requestMatchers(HttpMethod.GET, "favicon.ico")
                    .permitAll()
                    .requestMatchers(HttpMethod.GET, "/v3/api-docs/**")
                    .permitAll()
                    .requestMatchers(HttpMethod.GET, "/swagger-resources/**")
                    .permitAll()
                    .requestMatchers(HttpMethod.GET, "/swagger-ui/**")
                    .permitAll()
                    .requestMatchers(HttpMethod.GET, "/webjars/**")
                    .permitAll()
                    .requestMatchers(HttpMethod.GET, "/management/health")
                    .permitAll()
                    .requestMatchers(HttpMethod.GET, "/management/info")
                    .permitAll()
                    .requestMatchers(HttpMethod.GET, "/public/lists")
                    .permitAll()
                    .requestMatchers(HttpMethod.GET, "/public/lists/count")
                    .permitAll()
                    .anyRequest()
                    .authenticated());
    return http.build();
  }
}

File: src/main/kotlin/net/flyingfishflash/loremlist/core/configuration/WebSecurityConfiguration.kt
package net.flyingfishflash.loremlist.core.configuration

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.http.HttpMethod
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.annotation.web.invoke
import org.springframework.security.config.http.SessionCreationPolicy
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationConverter
import org.springframework.security.oauth2.server.resource.authentication.JwtGrantedAuthoritiesConverter
import org.springframework.security.web.SecurityFilterChain
import org.springframework.web.cors.CorsConfiguration
import org.springframework.web.cors.CorsConfigurationSource
import org.springframework.web.cors.UrlBasedCorsConfigurationSource

@Configuration
@EnableWebSecurity
class WebSecurityConfiguration {

  @Bean
  fun corsConfigurationSource(): CorsConfigurationSource {
    val configuration = CorsConfiguration()
    configuration.setAllowedOriginPatterns(listOf("*"))
    configuration.allowedMethods = listOf("*")
    configuration.allowedHeaders = listOf("*")
    configuration.allowCredentials = true
    configuration.exposedHeaders = listOf("X-Auth-Token", "Authorization")
    val source = UrlBasedCorsConfigurationSource()
    source.registerCorsConfiguration("/**", configuration)
    return source
  }

  @Bean
  fun jwtAuthenticationConverter(): JwtAuthenticationConverter {
    val grantedAuthoritiesConverter = JwtGrantedAuthoritiesConverter()
    grantedAuthoritiesConverter.setAuthoritiesClaimName("grants")

    val jwtAuthenticationConverter = JwtAuthenticationConverter()
    jwtAuthenticationConverter.setJwtGrantedAuthoritiesConverter(grantedAuthoritiesConverter)
    return jwtAuthenticationConverter
  }

  @Bean
  fun filterChain(http: HttpSecurity): SecurityFilterChain {
    http {
      cors { }
      csrf { disable() }
      oauth2ResourceServer {
        jwt { }
      }
      sessionManagement { sessionCreationPolicy = SessionCreationPolicy.STATELESS }
      authorizeHttpRequests {
        authorize(HttpMethod.GET, "favicon.ico", permitAll)
        authorize(HttpMethod.GET, "/v3/api-docs/**", permitAll)
        authorize(HttpMethod.GET, "/swagger-resources/**", permitAll)
        authorize(HttpMethod.GET, "/swagger-ui/**", permitAll)
        authorize(HttpMethod.GET, "/webjars/**", permitAll)
        authorize(HttpMethod.GET, "/management/health", permitAll)
        authorize(HttpMethod.GET, "/management/info", permitAll)
        authorize(HttpMethod.GET, "/public/lists", permitAll)
        authorize(HttpMethod.GET, "/public/lists/count", permitAll)
        authorize(anyRequest, authenticated)
//        hasAuthority("SCOPE_administrator")
      }
    }
    return http.build()
  }
}
